### PR TITLE
tree2: Inital recursive map and list support

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Maps and lists have some issues with directly recursive schema.

This change adds support for schema to opt out of implicit construction (so its valid to change their constructor types), then kludges the constructors of maps and lists so they are possible to use.

This is less than ideal, but may be able to be more easily cleaned up as simple-tree and class-tree are unified.

Since this is a hopefully temporary hack, the APIs for this are marked alpha not  beta.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/dds/tree2/src/test/class-tree/schemaFactory.examples.spec.ts [hunk 1]
```diff
@@ -113,4 +112,10 @@ describe("Class based end to end example", () => {
 		const theTree = factory.create(new MockFluidDataStoreRuntime(), "tree");
 		setup(theTree);
 	});
+
+	it("config2", () => {
```
Comment (Josmithr) [inline experimental/dds/tree2/src/test/class-tree/schemaFactory.examples.spec.ts:116]: A bit more description of what this is actually testing would be useful

File: experimental/dds/tree2/src/test/class-tree/schemaFactory.examples.spec.ts [hunk 2]
```diff
@@ -48,6 +48,8 @@ class Note extends schema.object("Note", {
 	}
 }
 
+type XX = typeof Note.implicitlyConstructable;
```
Comment (Josmithr) [inline experimental/dds/tree2/src/test/class-tree/schemaFactory.examples.spec.ts:51]: What is this for?

File: experimental/dds/tree2/src/test/class-tree/schemaFactory.recursiveExamples.spec.ts [hunk 3]
```diff
@@ -3,41 +3,108 @@
...
+			new TreeConfiguration(MapRecursive, () => new MapRecursive(undefined)),
+		);
+		const data = [...view.root];
+		assert.deepEqual(data, []);
+
+		// Nested
+		{
+			type T = InsertableTreeNodeFromImplicitAllowedTypes<typeof MapRecursive>;
+			const _check: T = new MapRecursive(undefined);
+			// Only explicitly constructed recursive lists are currently allowed:
```
Comment (Josmithr) [inline experimental/dds/tree2/src/test/class-tree/schemaFactory.recursiveExamples.spec.ts:102]:
```suggestion
			// Only explicitly constructed recursive maps are currently allowed:
```

File: experimental/dds/tree2/src/test/class-tree/schemaFactory.examples.spec.ts (hunk 1 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/class-tree/schemaFactory.examples.spec.ts:116]:
```suggestion
	// Confirm that the alternative syntax for the config from the example above (config2) actually works.
	it("config2", () => {
```

File: experimental/dds/tree2/src/test/class-tree/schemaFactory.examples.spec.ts (hunk 2 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/class-tree/schemaFactory.examples.spec.ts:51]: From my debugging. Shouldn't be there.
```suggestion
```

File: experimental/dds/tree2/src/shared-tree/sharedTree.ts [hunk 4]
```diff
@@ -298,7 +299,11 @@ export class SharedTree
 		nodeKeyManager?: NodeKeyManager,
 		nodeKeyFieldKey?: FieldKey,
 	): CheckoutFlexTreeView<TRoot> {
-		assert(this.hasView2 === false, 0x7f3 /* Cannot create second view from tree. */);
+		if (this.hasView2 === true) {
+			throw new UsageError(
```
Comment (Josmithr) [inline experimental/dds/tree2/src/shared-tree/sharedTree.ts:303]: Thanks!

File: experimental/dds/tree2/src/class-tree/schemaTypes.ts [hunk 5]
```diff
@@ -109,13 +112,25 @@ export interface TreeNodeSchemaClass<
...
+	out Kind extends NodeKind,
+	out ImplicitlyConstructable extends boolean,
 > {
 	readonly identifier: Name;
 	readonly kind: Kind;
-	readonly info: Specification;
+	readonly info: unknown;
+
+	/**
+	 * If true, when constructing insertable content,
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaTypes.ts:124]: Nit: this phrasing is a bit hard to parse. Might read a bit easier as something like "If true, the data that would have been passed to the node's constructor can be included instead of an unhydrated node when constructing insertable content."

File: experimental/dds/tree2/src/class-tree/schemaFactoryRecursive.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+			NodeKind.Map,
+			TreeMapNode<T>,
+			undefined,
+			false
+		>;
+	}
+}
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+export type Mappy<T extends ImplicitAllowedTypes> = {
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaFactoryRecursive.ts:99]: Are this and `Mappy2` needed? They're exported, but I don't see them used anywhere.

File: experimental/dds/tree2/src/class-tree/schemaFactoryRecursive.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+	TreeMapNode,
+	TreeNodeFromImplicitAllowedTypes,
+	TreeNodeSchemaClass,
+} from "./schemaTypes";
+import { SchemaFactory } from "./schemaFactory";
+
+/**
+ * Extends SchemaFactory with utilities for recursive types.
+ *
+ * This is separated from SchemaFactory as these APIs are more experimental, and may be stabilized independently.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaFactoryRecursive.ts:21]: Nit:
```suggestion
 * @remarks This is separated from SchemaFactory as these APIs are more experimental, and may be stabilized independently.
```

File: experimental/dds/tree2/src/class-tree/schemaFactory.ts [hunk 8]
```diff
@@ -530,10 +593,15 @@ export class SchemaFactory<TScope extends string, TName extends number | string
...
+	 * Running the command `TypeScript: Restart TS Server` with the schema file focused should fix it.
 	 * Sometimes this does not work: closing all open files except the schema before running the command can help.
 	 * Real compile errors (for example elsewhere in the file) can also cause the IntelliSense to not work correctly ever after `TypeScript: Restart TS Server`.
-	 * Intellisense has also shown problems when schema files with recursive types are part of a cyclic file dependency. Splitting the schema into its own file with minimal dependencies can help with this.
+	 *
+	 * Intellisense has also shown problems when schema files with recursive types are part of a cyclic file dependency.
+	 * Splitting the schema into its own file with minimal dependencies can help with this.
+	 *
+	 * Ensure `"noImplicitAny": true` is set in the `tsconfig.json`.
+	 * Without it recursive types that are not working properly can infer `any` and give very non-type-safe results instead of erroring.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaFactory.ts:604]: Nit
```suggestion
	 * Without it recursive types that are not working properly - they can infer `any` and give very non-type-safe results instead of erroring.
```

File: experimental/dds/tree2/src/class-tree/schemaFactory.ts [hunk 9]
```diff
@@ -323,28 +343,44 @@ export class SchemaFactory<TScope extends string, TName extends number | string
...
+				true
 			>;
 		}
-		return this.namedMap(nameOrAllowedTypes as TName, allowedTypes, true);
+		return this.namedMap(nameOrAllowedTypes as TName, allowedTypes, true, true);
 	}
 
-	private namedMap<Name extends TName | string, const T extends ImplicitAllowedTypes>(
+	/**
+	 * See note on `namedList`.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaFactory.ts:360]: Nit (just in case this class is ever exported):
```suggestion
	 * @privateRemarks See note on `namedList`.
```

**Review by Josmithr — COMMENTED** [review]
The changes look mostly good to me. My only remaining question is what `Mappy` and `Mappy2` are for. If they are necessary (and not just the remnants of testing), we should definitely document why.

File: experimental/dds/tree2/src/class-tree/schemaFactory.ts (hunk 9 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/class-tree/schemaFactory.ts:360]: This is a package public API which should be private, so making its only doc saying why its public private seems like a bad idea.

**Comment by CraigMacomber** [comment]
> The changes look mostly good to me. My only remaining question is what `Mappy` and `Mappy2` are for. If they are necessary (and not just the remnants of testing), we should definitely document why.

Left over from my many experiments to try and make maps work better (in some cases inlining random things fixed it). Those were not intended to be included in this version and have now been removed. Thanks for catching that!

File: experimental/dds/tree2/src/class-tree/schemaFactory.ts (hunk 8 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/class-tree/schemaFactory.ts:604]: I think you missed the "that" earlier in this sentence, as I think your change makes it malformed.

File: experimental/dds/tree2/src/class-tree/schemaTypes.ts [hunk 10]
```diff
@@ -109,13 +112,25 @@ export interface TreeNodeSchemaClass<
...
+	 * instead of including an {@link Unhydrated} node,
+	 * the data that would have been passed to the constructor can be included instead.
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/class-tree/schemaTypes.ts:126]:
```suggestion
	 * When constructing insertable content,
	 * data that could be passed to the node's constructor can be used instead of an {@link Unhydrated} node
	 * iff implicitlyConstructable is true.
```

File: experimental/dds/tree2/src/class-tree/schemaTypes.ts (hunk 5 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/class-tree/schemaTypes.ts]: fixed, I think.

File: experimental/dds/tree2/src/class-tree/schemaFactory.ts (hunk 9 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/class-tree/schemaFactory.ts:360]: Oh, I misread the situation. Since this is user-facing, it might be better to directly link to the referenced member (if we can), or duplicate the docs :/

File: experimental/dds/tree2/src/class-tree/schemaFactory.ts (hunk 8 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/class-tree/schemaFactory.ts:604]: Oh, yes you're right. So this would be more appropriate:
```suggestion
	 * Without it, recursive types that are not working properly can infer `any` and give very non-type-safe results instead of erroring.
```

File: experimental/dds/tree2/src/class-tree/schemaFactory.ts (hunk 9 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/class-tree/schemaFactory.ts:360]: Or something like :
```typescript
	 * Define a {@link TreeNodeSchema} for a {@link (TreeMapNode:interface)}.
	 *
	 * @param name - Unique identifier for this schema within this factory's scope.
	 *
	 * @remarks See remarks on {@link SchemaFactory.namedlist}.
```

**Review by Josmithr — APPROVED** [review]
Still a couple of docs suggestions, but otherwise looks good to me!

**Comment by CraigMacomber** [comment]
When switching from main to this branch, the incremental build (via `pnpm run build` in the tree2 package) with 33 errors. They start with:

```
@fluid-experimental/tree2: error during command 'tsc --project ./src/test/tsconfig.json' (exit code 2)
@fluid-experimental/tree2: dist/class-tree/schemaFactory.d.ts:118:97 - error TS2577: Return type annotation circularly references itself.
@fluid-experimental/tree2: 
@fluid-experimental/tree2: 118     map<Name extends TName, const T extends ImplicitAllowedTypes>(name: Name, allowedTypes: T): TreeNodeSchemaClass<`${TScope}.${Name}`, NodeKind.Map, TreeMapNode<T>, ReadonlyMap<string, InsertableTreeNodeFromImplicitAllowedTypes<T>>, true>;
```

And end with:
```
@fluid-experimental/tree2: src/test/class-tree/testRecursiveSchema.ts:44:14 - error TS2506: 'MapRecursive' is referenced directly or indirectly in its own base expression.
@fluid-experimental/tree2: 
@fluid-experimental/tree2: 44 export class MapRecursive extends sf.mapRecursive("Map", [MapRef2]) {}
@fluid-experimental/tree2:                 ~~~~~~~~~~~~
@fluid-experimental/tree2: 
@fluid-experimental/tree2: 
@fluid-experimental/tree2: Found 33 errors in 3 files.
@fluid-experimental/tree2: 
@fluid-experimental/tree2: Errors  Files
@fluid-experimental/tree2:      2  dist/class-tree/schemaFactory.d.ts:118
@fluid-experimental/tree2:     27  src/test/class-tree/schemaFactory.recursiveExamples.spec.ts:57
@fluid-experimental/tree2:      4  src/test/class-tree/testRecursiveSchema.ts:26
[7/8] x @fluid-experimental/tree2: tsc --project ./src/test/tsconfig.json - 3.652s
```

These can be fixed by running `pnpm run clean` in `experimental/dds/tree2`.
